metric_name_builder_bench_test.go

```diff
@@ -0,0 +1,136 @@
+package prometheus
+
+import (
+	"fmt"
+	"testing"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+func BenchmarkBuildMetricNames(b *testing.B) {
+	metrics := createTestMetrics()
+
+	for _, workerType := range []string{"BuildCompliantMetricName", "BuildMetricName"} {
+		for _, withSuffixes := range []bool{true, false} {
+			b.Run(fmt.Sprintf("%s/withSuffixes=%t", workerType, withSuffixes), func(b *testing.B) {
+				var worker Worker
+				if workerType == "BuildCompliantMetricName" {
+					worker = &BuildCompliantMetricNameWorker{
+						metrics:      metrics,
+						namespace:    "test_namespace",
+						withSuffixes: withSuffixes,
+					}
+				} else {
+					worker = &BuildMetricNameWorker{
+						metrics:      metrics,
+						namespace:    "test_namespace",
+						withSuffixes: withSuffixes,
+					}
+				}
+
+				b.ResetTimer()
+				for i := 0; i < b.N; i++ {
+					worker.Work()
```

Comment:
The worker type looks like over-engineering to me, and just makes the benchmark more difficult to comprehend and maintain. Can't you just replace with the following?

```suggestion
					fn := BuildMetricName
					if funcName == "BuildCompliantMetricName" {
						fn = BuildCompliantMetricName
					}
					for _, metric := range metrics {
							fn(metric, "test_namespace", withSuffixes)
					}
```

Reply:
Looping over metrics looks wrong to me as well. You should instead make a sub-benchmark per metric, so one can see the performance per case.

Reply:
Makes the benchmark a lot slower, but it should be fine

Reply:
> Makes the benchmark a lot slower, but it should be fine

Benchmarks are generally slow. The quality of a benchmark is measured by the information you can get out of it, not how fast it runs.